### PR TITLE
Merge axum branch into main

### PR DESCRIPTION
**Pull request by joeftiger**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,20 +1,32 @@
...
 [dev-dependencies]
-serde_json = "1.0"
-serde_urlencoded = "0.7"
+serde_json = "1.0.117"
+serde_urlencoded = "0.7.1"
+
+[profile.release]
+codegen-units = 1
+lto = "fat"
+opt-level = 3
```
Comment (mTsBucy1) [inline Cargo.toml]:
```suggestion
opt-level = 3

```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+# Copy binary to /bin/server for future use
+RUN mkdir /app/bin
+RUN cp target/$PROFILE/rusty-runner-server /app/bin/server
+# Copy healthcheck binary
+COPY --from=httpget /httpget /app/bin/httpget
+
+FROM rust:slim AS runtime
+LABEL name="rusty-remote-runner" \
+      maintainer="info@meintest.software" \
```
Comment (mTsBucy1) [inline Dockerfile]: should maybe be developers@...?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# syntax=docker/dockerfile:1
+
+# This dockerfile builds the rusty-runner-server service as a container.
+# The service starts automatically when running the container on
+# - host: 127.0.0.1
+# - port: 8000
```
Comment (mTsBucy1) [inline Dockerfile]: The host is mostly irrelevant and you override it anyway

```suggestion
# The service starts automatically when running the container on port 8000
```

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    --no-create-home \
+    --uid "${UID}" \
+    appuser
+USER appuser
+
+COPY --from=builder /app/bin /bin
+
+EXPOSE 8000
+HEALTHCHECK --interval=30s --timeout=3s --retries=3 CMD ["/bin/httpget", "http://127.0.0.1:8000/health"]
+ENTRYPOINT ["/bin/server", "--host=0.0.0.0"]
```
Comment (mTsBucy1) [inline Dockerfile]:
```suggestion
ENTRYPOINT ["/bin/server", "--host=0.0.0.0"]

```

File: rusty-runner-server/Cargo.toml [hunk 5]
```diff
@@ -1,16 +1,30 @@
...
+env_logger = { version = "0.11.3", features = ["unstable-kv"] }
+# -- Util
+fastrand = "2.1.0"
+# -- CLI
+clap = { version = "4.5.4", features = ["derive", "env"] }
+
+[dev-dependencies]
+anyhow = "1.0.86"
+httpc-test = "0.1.9"
+serde_json = "1.0.117"
```
Comment (mTsBucy1) [inline rusty-runner-server/Cargo.toml]:
```suggestion
serde_json = "1.0.117"

```

File: rusty-runner-server/src/main.rs [hunk 6]
```diff
@@ -1,37 +1,101 @@
...
+        .init();
+
+    let CliArgs { host, port } = CliArgs::parse();
 
     log::info!(
-        "Starting server v{} [api v{}]",
-        env!("CARGO_PKG_VERSION"),
-        rusty_runner_api::api::VERSION
+        version = env!("CARGO_PKG_VERSION"),
+        api_version =rusty_runner_api::api::VERSION;
```
Comment (mTsBucy1) [inline rusty-runner-server/src/main.rs]:
```suggestion
        api_version = rusty_runner_api::api::VERSION;
```

File: rusty-runner-server/src/routes.rs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    Json(InfoResponse {
+        api_version: String::from(VERSION),
+        #[cfg(windows)]
+        computer_name: std::env::var("COMPUTERNAME").unwrap_or(String::from("{unknown}")),
+        #[cfg(windows)]
+        os_type: OsType::Windows,
+        #[cfg(unix)]
+        computer_name: std::env::var("HOSTNAME").unwrap_or(String::from("{unknown}")),
+        #[cfg(unix)]
+        os_type: OsType::Unix,
```
Comment (mTsBucy1) [inline rusty-runner-server/src/routes.rs]: All these configs could lead to weird errors if building without exactly one of them set, and you removed by checks for that.

File: rusty-runner-server/src/main.rs [hunk 8]
```diff
@@ -1,37 +1,101 @@
...
+        .route("/health", get(|| async { "OK" }))
+        .layer(TraceLayer::new_for_http());
+
+    let addr = format!("{host}:{port}");
+    let listener = tokio::net::TcpListener::bind(addr).await?;
+    log::info!(
+        addr:display = host,
+        port = port;
+        "listening to TCP"
+    );
```
Comment (mTsBucy1) [inline rusty-runner-server/src/main.rs]: Nitpick you can ignore, but I have made a change to this pattern in the template, to not need to format and parse the address, and allow for auto-binding of a port (only really useful for manual experimentation)

File: rusty-runner-server/src/routes.rs [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    fn route_logged(self, path: &str, method_router: MethodRouter) -> Self;
+    /// [`Router::nest_service`] and log which nested service is being added to the router.
+    fn nest_service_logged(self, path: &str, method_router: MethodRouter) -> Self;
+}
+
+impl LoggedExt for Router {
+    fn route_logged(self, path: &str, method_router: MethodRouter) -> Self {
+        log::info!(path; "adding sub-route");
+        self.route(path, method_router)
+    }
```
Comment (mTsBucy1) [inline rusty-runner-server/src/routes.rs]: I'm not too sure about this, it seems confusing. I guess the idea is that you can see the "API docs" when starting the server?
Then firstly, the wording is confusing: what does it mean to "add" a route, and why a sub-route. Nitpicky opinion, but I think with this wording, it is at best a debug level, not info.
For an info, you would need to log the full path and preferrably also the explicitly state the method (GET) and word it in a static way, i.e. "available route", but at this point you are just printing the api docs, which should more usefully be put into the CLIArgs help text.

File: rusty-runner-server/src/service.rs [hunk 10]
```diff
@@ -1,116 +0,0 @@
...
-
-use actix_web::{get, post, web, HttpResponse};
-use tokio::process::Command;
-
-use crate::process::{process_command, working_directory};
-
-#[cfg(all(windows, unix))]
-compile_error!("Unix and Windows are exclusive!");
-#[cfg(not(any(windows, unix)))]
-compile_error!("Either Unix or Windows must be targeted!");
```
Comment (mTsBucy1) [inline rusty-runner-server/src/service.rs]: This should not have been removed

File: src/api.rs [hunk 11]
```diff
@@ -192,6 +172,7 @@ pub enum RunStatus {
         exit_code: i32,
         /// If `return_logs` is set, this returns a tuple of the raw `stdout` and `stderr`
         /// logs.
+        // TODO: It would be MUCH better for consumer clients if this was split into 2 fields.
         std_out_and_err: Option<(Vec<u8>, Vec<u8>)>,
```
Comment (mTsBucy1) [inline src/api.rs]: This is a breaking change, so I suggest you do this now, instead of provoking another one later on :)

File: src/lib.rs [hunk 12]
```diff
@@ -11,19 +11,19 @@
 //! * `GET /api/info` returns an informative [`api::InfoResponse`] object.
```
Comment (mTsBucy1) [inline src/lib.rs]: If you are not adding `/health` here to the API, then you should not call it to check server availability as is currently done in `runner-router`. Altho I would only use `/api/info` anyway, as it is backwards compatible (at least to tell you that it is incompatible).

File: src/lib.rs [hunk 13]
```diff
@@ -11,19 +11,19 @@
...
 //! * `POST /api/run` runs a command analogous to [`std::process::Command`].
 //! * `POST /api/runscript` runs a script with a given interpreter.
-//! * `GET /api/file` fetches a file from the servers file system.
+//! * `GET /api/file/{path}` fetches a file from the servers file system.
 //!
 //! ## Working with files
 //! The working directory of the executed commands is implementation defined,
 //! but the same for all methods and constant over the lifetime of the server.
-//! The path for file fetching is also relative to this directory.
+//! The path for file fetching is also relative to (and confined to!) this directory.
```
Comment (mTsBucy1) [inline src/lib.rs]:
```suggestion
//! The path for file fetching is also a relative path in this directory.
```

File: src/lib.rs [hunk 14]
```diff
@@ -11,19 +11,19 @@
...
 //!
 //! Best use a relative randomly named subdirectory for your file operations.
 //! E.g. `./task-9ae4ef2b9d13/your-file`
 //!
 //! ## Long running jobs
 //! Using `reqwest` and `actix_web` does not impose and significant timeout on
-//! the rest calls. Therefore currently the rest calls will just wait untill
+//! the rest calls. Therefore currently the rest calls will just wait until
 //! the command terminates and return then.
```
Comment (mTsBucy1) [inline src/lib.rs]: Outdated information. Check again for axum (there is a test at `taumada\crates\services\runner-router-server\src\model\runner\mod.rs:test_http_timeout` to help you)

**Review by mTsBucy1 — CHANGES_REQUESTED** [review]
Nice. I have left some comments, some more serious, others more nitpicky.

Afterwards, I don't think a rebase is necessary, we'll just force-push the main branch, there are only two commits there, and one was my fault for not seeing your changes, the other is already incorporated anyway.

**Comment by mTsBucy1** [comment]
And make sure the readme is up to date with the rustdoc. Either run `cargo readme`, or I just found about about this neat piece of code: `#![doc = include_str!("../README.md")]`

File: rusty-runner-server/src/routes.rs (hunk 7 above)
Comment (joeftiger) [reply rusty-runner-server/src/routes.rs]: What do you mean by "you removed by checks for that."?
I copied this code snippet from the previous actix implementation here:
https://github.com/meinTest/rusty-remote-runner/blob/main/rusty-runner-server/src/service.rs#L17-L32

File: src/lib.rs (hunk 13 above)
Comment (joeftiger) [reply src/lib.rs]: I'd rather be explicit in that the path is confined to that directory.

File: rusty-runner-server/src/routes.rs (hunk 7 above)
Comment (mTsBucy1) [reply rusty-runner-server/src/routes.rs]: "my check", typo. [see ](https://github.com/meinTest/rusty-remote-runner/pull/2#discussion_r1741528243)

Comment (joeftiger) [reply rusty-runner-server/src/routes.rs]: We could use https://docs.rs/gethostname instead. This code block here is the only reason why would have to enforce unix / windows checks.

File: rusty-runner-server/src/service.rs (hunk 10 above)
Comment (joeftiger) [reply rusty-runner-server/src/service.rs]: See https://github.com/meinTest/rusty-remote-runner/pull/2#discussion_r1743346155
And I do not see how one can compile for windows/unix at the same time?

File: rusty-runner-server/src/routes.rs (hunk 7 above)
Comment (mTsBucy1) [reply rusty-runner-server/src/routes.rs]: Nope, https://github.com/meinTest/rusty-remote-runner/pull/2/files#diff-ec2e0ddaf5d7247ad9b34ec3ad5dd697b19114e2eff687bba6b013e3a842dbf0R80-R84 also uses the pattern. I don't care about only supporting unix / windows, that is perfectly fine, just don't remove checks without fixing the issue.

File: rusty-runner-server/src/service.rs (hunk 10 above)
Comment (mTsBucy1) [reply rusty-runner-server/src/service.rs]: Doesn't matter if it is dead code, it is a sanity check.

File: src/lib.rs (hunk 13 above)
Comment (mTsBucy1) [reply src/lib.rs]: Yeah, but the exclaimation mark and parathesis are too strong of a statement. Make it "a relative path inside this directory", then it is even clearer without making the wrong point. The point isn't that you shouldn't be able to access anything else, hell this is RCEaaS, it is that you know wether to give a full path or a just the relative part.

File: rusty-runner-server/src/routes.rs (hunk 7 above)
Comment (joeftiger) [reply rusty-runner-server/src/routes.rs]: > I don't care

k

**Pull request merged**